### PR TITLE
Tigers - Melley Gebretatios

### PR DESCRIPTION
**Pull request by mell2022**

### CODE REVIEW

**Comment by kathawkins** [comment]
> 

Adagrams project for Melley G. and Kat H.

File: adagrams/game.py [hunk 1]
```diff
@@ -1,11 +1,124 @@
...
+    
+    #Appends 10 random letters to the hand
+    for _ in range(10):
+        current_letter = random.choice(adjusted_letter_pool)
+        adjusted_letter_pool.remove(current_letter)
+        hand_list.append(current_letter)
+
+    return hand_list
+
+def uppercase_letter_list(letter_bank):
```
Comment (spitsfire) [inline adagrams/game.py:42]: Interesting check! If this game were made more flexible, like allowing a user to pass in their own letter bank, it might be a good idea to have this!

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,124 @@
...
+    hand=[]
+
+    #Converts each letter in the letter bank to uppercase
+    for letter in letter_bank:
+        upper_letter = letter.upper()
+        hand.append(upper_letter)
+
+    return hand
 
 def uses_available_letters(word, letter_bank):
```
Comment (spitsfire) [inline adagrams/game.py:52]: 👍

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,124 @@
...
+    hand = uppercase_letter_list(letter_bank)
+    hand_dict = {}
+
+    #Generates a counter dictionary for letters in the hand
+    for hand_letter in hand:
+        # lowercase_hand_letter = hand_letter.upper()
+        if hand_letter in hand_dict:
+            hand_dict[hand_letter] += 1
+        else:
+            hand_dict[hand_letter] = 1
```
Comment (spitsfire) [inline adagrams/game.py:63]: Great job using a frequency map, or dictionary, to handle this. Dictionaries have a faster lookup time than lists when used with `if...in`.

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,124 @@
...
+        if hand_dict[player_letter] == 0:
+            return False
+        else:
+            hand_dict[player_letter] -= 1
+    #Assumes a valid word if the function does not return False
+    return True
+
+def score_word(word):    
+    players_letters_list=list(word.upper())
+    letter_count=len(players_letters_list)
```
Comment (spitsfire) [inline adagrams/game.py:79]: Let's get rid of this! Since we only care about the length one time in our function, let's calculate it right there.

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,124 @@
...
+    players_word_score = 0
+
+    #Adds each letter's points to the word score 
+    for player_letter in players_letters_list:
+        players_word_score += SCORE_DICT[player_letter]
+    
+    #Adds 8 points to the word score if the word is 7 letters or longer
+    if letter_count >= 7:
```
Comment (spitsfire) [inline adagrams/game.py:87]:
```python
if len(letter_count) > 6:
    # the rest of your code here
```

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,124 @@
...
+        current_score = score_word(word)
+        if current_score > max_score:
+            max_score = current_score
+            max_words = [word]
+            letter_counts = [len(word)]
+        elif current_score == max_score:
+            max_words.append(word)
+            letter_counts.append(len(word))
+
+    min_letter_count = min(letter_counts)
```
Comment (spitsfire) [inline adagrams/game.py:111]: Hmm the logic here is a little hard to follow. We have two lists that are keeping track of values that relate to each other. That's a hint that we should use a dictionary to carry the word and its length together as key-value pairs.

I think, however, we can do all of this in one for loop without creating two more below, which iterate over the same list twice. It also gets rid of the two lists `max_words` and `letter_counts`.

Maybe something like:
```python
for word in word_list:
    current_score = score_word(word)
    if current_score > max_score:
        # assign the current word to a max_word variable
        # assign the current score to a max_score variable
    elif current_score == max_score:
        # now bring in the logic inside the two for loops below
        # is the current word 10 letters long and longer than the max_word? 
            # it's now the max_word
        # is the current word shorter than the max_word?
            # it's now the max_word
```

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,124 @@
...
+'S': 1,
+'T': 1, 
+'U': 1, 
+'V': 4, 
+'W': 4, 
+'X': 8, 
+'Y': 4, 
+'Z': 10}
+
 def draw_letters():
```
Comment (spitsfire) [inline adagrams/game.py:30]: 👍

**Review by spitsfire — COMMENTED** [review]
Great job working together, Melley and Kat!

Here are a few things to consider for future projects that I've detailed below in your code:
- Consider refactoring the logic inside `get_highest_score` function. It was hard to follow, unlike the rest of your code. There are a lot of extra variables and repetitive for loops that iterate over the same lists.

If you have any questions about the feedback, please reach out to me!

File: adagrams/game.py [hunk 8]
```diff
@@ -1,11 +1,124 @@
+import random
+LETTER_POOL = list("A"*9 + "B"*2 + "C"*2 + "D"*4 + "E"*12 + "F"*2 + "G"*3 + "H"*2 + "I"*9 + "J"*1 + "K"*1 + "L"*4 + "M"*2 + "N"*6 + "O"*8 + "P"*2 + "Q"*1 + "R"*6 + "S"*4 + "T"*6 + "U"*4 + "V"*2 + "W"*2 + "X"*1 + "Y"*2 + "Z"*1)
```
Comment (spitsfire) [inline adagrams/game.py:2]: Creative approach! This will allow for a better distribution of probability when picking letters randomly.
